src/components/FacetController.js

```diff
@@ -23,202 +30,147 @@ function validFrameProps(originalProps, frameType) {
-              additionalSettings.xExtent = calculatedXExtent
-            }
-            if (calculatedYExtent) {
-              additionalSettings.yExtent = calculatedYExtent
-            }
-
-            return React.cloneElement(child, additionalSettings)
-          })
-        })
-      })
+    this.state = {
+      extents: {},
+      facetHover: undefined
     }
   }
 
-  mapChildrenWithAppropriateProps = ({
-    child,
-    facetProps,
-    rExtentPromises,
-    xExtentPromises,
-    yExtentPromises
-  }) => {
-    const rest = { ...facetProps }
-    const frameType = child.type.displayName
-
-    if (
-      facetProps.hoverAnnotation === true ||
-      facetProps.pieceHoverAnnotation === true
-    ) {
-      rest.customHoverBehavior = d => this.setState({ facetHover: d })
-    }
+  /**
+   * Helper for creating extent if we have a  min/max value
+   * use that else use the onChange version so we can in return
+   * normalize all of the facets to have the same extents
+   */
+  createExtent = (extentType: string, state: State) => {
+    return state.extents && state.extents[extentType]
+      ? state.extents[extentType]
+      : { onChange: this.extentHandler(extentType) }
+  }
 
-    if (rExtentPromises && frameType === "OrdinalFrame") {
-      let extentCallback
-      const extentPromise = new Promise((res, rej) => {
-        extentCallback = d => {
-          if (d) {
-            res(d)
-          } else {
-            rej("rExtent calculation failed")
+  /**
+   * Whenever the extent changes, create the min/max values for each extentType
+   * so this could be rExtent for OrdinalFrame or x/yExtent for the XYFrame
+   */
+  extentHandler = (extentType: string) => {
+    return (extentValue: Array<number>) => {
+      this.setState(prevState => {
+        const extentMinMaxValues = (prevState.extents[extentType] || []).concat(
+          extentValue
+        )
+
+        return {
+          extents: {
+            ...prevState.extents,
+            [extentType]: [
+              Math.min(...extentMinMaxValues),
+              Math.max(...extentMinMaxValues)
+            ]
           }
         }
       })
 
-      rExtentPromises.push(extentPromise)
-      rest.rExtent = { onChange: extentCallback }
+      return extentValue
     }
+  }
 
-    if (xExtentPromises && frameType === "XYFrame") {
-      let extentCallback
-      const extentPromise = new Promise((res, rej) => {
-        extentCallback = d => {
-          if (d) {
-            res(d)
-          } else {
-            rej("xExtent calculation failed")
-          }
-        }
-      })
-
-      xExtentPromises.push(extentPromise)
-      rest.xExtent = { onChange: extentCallback }
-    }
+  /**
+   * Remove and add required annotation props for specific annotation types.
+   */
+  mapChildrenWithHoverAnnotations = ({
+    child,
+    state
+  }: {
+    child: Element<*>,
+    state: State
+  }) => {
+    const annotations = child.props.annotations || []
```

Comment:
After my last commit, annotations are no longer being removed on hover due to this is only additive. This was previously broken in the original PR. Should the `child.props.annotations` be used in this case since it currently only supports `hoverAnnotation`?